stdlib.sh

```diff
@@ -654,12 +840,443 @@ function __STDLIB_API_1_info() { # {{{
 # This function may be overridden
 #
 function __STDLIB_API_1_debug() { # {{{
-	(( std_DEBUG )) && __STDLIB_API_1_std::log >&2 "DEBUG: " "${*:-Unspecified message}"
+	local message="${1:-Unspecified message}"
+	local -i level=${2:-$STDLIB_DEBUG_NORMAL}
+
+	(( $std_DEBUG == $level )) && __STDLIB_API_1_std::log >&2 "[  ${std_COLOUR_START_DEBUG}DEBUG${std_COLOUR_END}  ]" "${message}"
 
 	# Don't stomp on std_ERRNO
 	return $(( ! std_DEBUG ))
 } # __STDLIB_API_1_debug # }}}
 
+###############################################################################
+#
+# stdlib.sh - Command execution
+#
+# The following set of function are meant to ease usage of control variables
+# like DEBUG and PRETEND over sub commands, preserving preamble and output text.
+#
+# - The command being executed is stored in STDLIB_EXEC_COMMAND.
+# - Execution status code is always captured, saved in STDLIB_EXEC_STATUS and
+#   then evaluated.
+# - When the function return value is required, then the command is evaluated in
+#   a subshell and the execution return value is stored in STDLIB_EXEC_RESULT
+#
+# The default behaviour is to always use a subshell, display command's stdout
+# content, store both status and return value, not display command result.
+# This behaviour can be controlled by use of 'std_STREAM_XXX' parameter, in
+# order to be able to execute commands from completely silent output to full
+# interactive mode, while still being able to use the output control framework
+# provided by the function set.
+#
+# A number of support functions is provided for special handling of results and
+# display control.
+#
+# Typical usage should look as follows:
+#
+# local dir="/usr/local/bin"
+# waitForRes "Chdir to '${dir}'"
+# evalRes "cd ${dir}"
+# checkRes f "cannot chdir to '${dir}'"
+#
+# Normal output would be something like:
+#
+# [ EXEC  ] Chdir to '/usr/local/bin' [OK]
+#
+# If called with DEBUG=1, the output would be:
+#
+# [ EXEC  ] Chdir to '/usr/local/bin'
+# [ DEBUG ] About to run: 'cd /usr/local/bin'
+#  [OK]
+#
+# If called with PRETEND=1, the output would be:
+#
+# [ EXEC  ] Chdir to '/usr/local/bin'
+# [ DEBUG ] Pretend to run: 'cd /usr/local/bin'
+#  [OK]
+#
+###############################################################################
+
+# This function may be overridden
+#
+# The idea is to log a message (with optional new line at the end),
+# wait for exectuion to finish, then
+# log the overall result at the end of the same line
+# with one of the closing-tag function below.
+function __STDLIB_API_1_waitForRes() { # {{{
```

Comment:
Admittedly these functions could use a name change: I ended up porting them as-is from one of my previous implementations that was the result of many changes, and where I could not afford to break back compatibility.

The suggested 'execute' function is OK, but you still want to be able to call the three existing subfunctions in order to allow for interleaved code to be executed (like additional debugging, utility function calls, custom output and so on).